packet/packet.go

```diff
@@ -686,11 +686,91 @@ func InitEmptyEtherIPv6UDPPacket(packet *Packet, plSize int) {
 	packet.UDP.DgramLen = uint16(UDPLen + plSize)
 }
 
-// Swapping bytes for protocol numbers in Little Endian and Big Endian
+// Swapping uint16 in Little Endian and Big Endian
 func SwapBytesUint16(x uint16) uint16 {
 	return x<<8 | x>>8
 }
 
-func GetRawPacketBytes(pkt *Packet) []byte {
-	return low.GetRawPacketBytesMbuf(pkt.CMbuf)
+// Swapping uin32 in Little Endian and Big Endian
+func SwapBytesUint32(x uint32) uint32 {
+	return ((x & 0x000000ff) << 24) | ((x & 0x0000ff00) << 8) | ((x & 0x00ff0000) >> 8) | ((x & 0xff000000) >> 24)
+}
+
+// GetRawPacketBytes returns all bytes from this packet. Not zero-copy.
+func (packet *Packet) GetRawPacketBytes() []byte {
+	return low.GetRawPacketBytesMbuf(packet.CMbuf)
+}
+
+// GetPacketLen returns length of this packet
+func (packet *Packet) GetPacketLen() uint {
+	return low.GetDataLenMbuf(packet.CMbuf)
+}
+
+// EncapsulateHead adds bytes to packet. start - number of beginning byte, length - number of
+// added bytes. This function should be used to add bytes to the first half
+// of packet. Return false if error.
+func (packet *Packet) EncapsulateHead(start uint, length uint) bool {
+	if low.PrependMbuf(packet.CMbuf, length) == false {
+		return false
+	}
+	packet.Unparsed -= uintptr(length)
+	for i := uint(0); i < start; i++ {
+		*(*uint8)(unsafe.Pointer(packet.Unparsed + uintptr(i))) = *(*uint8)(unsafe.Pointer(packet.Unparsed + uintptr(i+length)))
+	}
+	return true
+}
+
+// EncapsulateTail adds bytes to packet. start - number of beginning byte, length - number of
+// added bytes. This function should be used to add bytes to the second half
+// of packet. Return false if error.
+func (packet *Packet) EncapsulateTail(start uint, length uint) bool {
+	if low.AppendMbuf(packet.CMbuf, length) == false {
+		return false
+	}
+	packetLength := packet.GetPacketLen()
+	for i := packetLength - 1; int(i)-int(length) >= int(start); i-- {
```

Comment:
Possibly remove condition calculation by changing loop condition to int(i) >= int(start)+int(length) because int(start)+int(length) may be calculated just once while int(i)-int(length) has to be calculated on every loop iteration.